### PR TITLE
Add a Vulkan HW interop

### PR DESCRIPTION
**Pull request by philipl**
This set of changes adds an HW interop (hwdec in our terminology, but there's no decoding going on here) to allow Vulkan frames produced by ffmpeg filters (and eventually vulkan video decoding) to be consumed directly by vo_gpu and vo_gpu_next without any reuploading or copying etc.

Pre-final Vulkan hwaccel support
--------------------------------------------
- [x] Pass full ra_ctx to hwdecs - needed to fully initialise the Vulkan hwdevice
- [x] Basic implementation of the HW interop
- [x] Replace use of the libplacebo `AVFrame` based wrapper with code that directly works with the `mp_image`
- [x] Fix validation errors due to not syncing layout when returning frames to ffmpeg
- [x] Implement a synchronised wrapper for VkQueue operations as we issue commands from both decode and vo threads
- [x] Provide a way to hwupload existing hw frames (cuda/vaapi) to Vulkan. We aren't smart enough to do this today
- [x] Don't require libplacebo v4.200+.
- [x] Add waf support
- [x] Add the right conditionals so the code builds with older ffmpeg and Vulkan SDK
- [x] Add hw->hw upload support to `format` filter
- [x] Add hw->hw map support to `format` filter

Final Vulkan hwaccel support
---------------------------------------
- [x] Require libplacebo v5.274+ for locking primitives, feature detection, and decode queue initialisation
- [x] Use ffmpeg queue and frame locking
- [x] Enable all necessary features and extensions for video decoding and new ffmpeg filters 
- [x] Support multiplane images
- [x] ffmpeg check based on checking for a #define. Can detect the Vulkan branch for now.
- [ ] Add the right conditional checks for sufficiently new versions of ffmpeg after the final hwaccel is merged

Testing so far:
- nvidia works for both hardware decode and nvdec hwupload
- AMD works for both hardware decode and vaapi hwmap
- Intel works for vaapi hwmap and h.264 decode (h.265 still pending)
- Intel and AMD work for softwware hwupload with `--vd-lavc-dr=no`
- nvidia and AMD work with filters with various caveats due to driver limitations
- Intel doesn't support descriptor buffers so no filters work

nvidia work-arounds
---------------------------

People trying to use non-vulkan-beta nvidia drivers will need my ffmpeg branch with some nvidia specific work-arounds that aren't in the main vulkan branch.

https://github.com/philipl/FFmpeg/tree/vulkan-nvidia-workarounds

### CODE REVIEW

**Comment by philipl** [comment]
Example command line:

```
mpv --hwdec=no --gpu-hwdec-interop=vulkan --vf=hwupload,vflip_vulkan
```

Cannot be used with hwdec until I put in better hwdevice selection.

File: video/out/vulkan/utils.c
Comment (haasn) [inline video/out/vulkan/utils.c:6]: `PTHREAD_MUTEX_INITIALIZER`

File: video/out/vulkan/utils.c
Comment (haasn) [inline video/out/vulkan/utils.c:75]: This would normally be a device-level command, not an instance-level command

File: video/out/vulkan/utils.c
Comment (haasn) [inline video/out/vulkan/utils.c:54]: To clarify what I mean further below: this doesn't work. There can be multiple `VkDevice` per `VkInstance`. The second invocation would override the first invocation's `vk_queue_submit` pointer, and end up with humbug.

File: video/out/vulkan/utils.c
Comment (haasn) [reply video/out/vulkan/utils.c:67]: > Technically this function also needs to be able to serve all device-level functions, in which case it has to maintain a per-device dispatch layer.

Rather, the other way around - we need the per-device dispatch layer for the *device* functions proc addr.

File: video/out/vulkan/utils.c
Comment (haasn) [reply video/out/vulkan/utils.c:54]: Possible work-around: use `vkGetInstanceProcAddr` instead of `vk_device_get_proc_addr` for these functions. That way it goes through the loader's own dispatch layer. That's, of course, one function call less efficient, but I doubt anybody cares here.

**Comment by haasn** [comment]
FFmpeg is in the process of gaining lock_queue/unlock_queue functions which will solve the queue threading issue (*), in addition to lock_frame/unlock_frame (which we will need to put inside map/unmap respectively)

(*) We will need a new libplacebo API though to forward these to the imported vulkan context

**Comment by cyanreg** [comment]
I know you're in a rush to merge this, and sorry you had to spend so much work, but could you wait a day or two so I can merge what's needed in ffmpeg? That way you wouldn't have to maintain that less than nice code.

**Comment by philipl** [comment]
Sure, I can wait. I still need to sort out the remaining open items. Thanks for your work on the ffmpeg side!

**Comment by Obegg** [comment]
This PR close this issue: 
https://github.com/mpv-player/mpv/issues/9675

**Comment by cyanreg** [comment]
No, it doesn't.

**Comment by Obegg** [comment]
> No, it doesn't.

The issue is a request for a vulkan hwdec, this is the PR that adds it.

**Comment by haasn** [comment]
> > No, it doesn't.
> 
> The issue is a request for a vulkan hwdec, this is the PR that adds it.

Don't we still need extra code to hook up the `--hwdec` option?

**Comment by philipl** [comment]
There are at least four things we need for working vulkan video decode acceeleration
- Drivers that actually implement it. nvidia hasn't included it in any non-developer driver release and the mesa driver work has only just begun
- ffmpeg hwaccels. @cyanreg has been working on this and it's close to done
- The hw interop that can pass VkImages to the vo without conversions. That's this change
- libplacebo changes to ensure we request whatever the necessary video extensions are and set up the video queues

**Comment by philipl** [comment]
I've done a bunch of refinement to the existing patch set and I'm pretty satisfied with it. The main thing right now is ideally I wouldn't push the ugly wrapper code for locking VkQueue access, and wait for the lock/unlock helpers to land in ffmpeg and libplacebo.

The ra and filter changes can land before the actual vulkan interop, and maybe that's the best way to move things forward.

**Comment by philipl** [comment]
Added HW->HW upload support to the `format` filter so now you can do:

```
mpv --hwdec=nvdec --vf=format=fmt=vulkan,vflip_vulkan
```

**Comment by philipl** [comment]
Ok, hwmap support in `vf_format` is now present.

So you can do

```
mpv --hwdec=vaapi --vf=format=fmt=vulkan,vflip_vulkan
```

on Intel, and presumably AMD.

**Review by laichiaheng — APPROVED** [review]

**Comment by GreyXor** [comment]
thanks,
Is there any need to test things here?

**Comment by philipl** [comment]
You can try it out if you want. I don't have AMD hardware so testing there is always useful.

But right now we're blocked on getting locking primitives into ffmpeg. @cyanreg has been working on that.

**Comment by GreyXor** [comment]
> 

Yep i have AMD iGPU vega 8, i'll test that.

**Comment by philipl** [comment]
Updated to use the new locking primitives. These have been upstream in libplacebo for a while but the ffmpeg work is still not merged, so I cannot do a correct libavutil version check yet.

Note that the ffmpeg code that is going to be merged uses multi-plane images which we cannot currently handle. That requires both WIP changes in libplacebo and a bunch of work in mpv that I've only scratched the surface of (and is not in this series).

I did all my testing with single plane formats (eg: rgba gifs and 12bit 444 content decoded by Intel vaapi to xv36)

This also means cuda vulkan interop doesn't currently work as cuda cannot handle multi-plane images and the new hwcontext_vulkan code doesn't (has lost the ability to?) work with multi-image based frames.

@cyanreg @haasn

**Comment by philipl** [comment]
I have updated the patch set to disable multiple plane for now. Until we support rendering multiplane images, we can't actually support the vulkan decoder, but we can support uploaded images as in my original patchset, so we disable multiplane for that purpose.

Supporting filters requires a very new vulkan extension, which only radv currently supports.

**Comment by philipl** [comment]
Indeed. I've tested it and it works.

**Comment by haasn** [comment]
Libplacebo master supports multiplane images

**Comment by philipl** [comment]
General status update:
- Multiplane is working great
- Main remaining blockers are:
  - Testing
  - Merging video queue configuration in libplacebo
  - Coming up with a way to test for vulkan features to fail gracefully

**Comment by Obegg** [comment]
> ffmpeg filters (and eventually vulkan video decoding) to be consumed directly by vo_gpu and vo_gpu_next without any reuploading or copying etc.

If I understand correctly - because there is no reuploading or copying etc it's better than any other interop, 
So should it be the default interop? (I guess not initially in fear of unexpected bugs and other issues that may rise such as users without the correct driver and etc, but surely this would be the default interop at some point?)

**Comment by philipl** [comment]
> If I understand correctly - because there is no reuploading or copying etc it's better than any other interop, So should it be the default interop? (I guess not initially in fear of unexpected bugs and other issues that may rise such as users without the correct driver and etc, but surely this would be the default interop at some point?)

I'm sure there's a point in the future where you'd prefer `vulkan` when doing `hwdec=auto` but that's a good way down the line. It's unclear how well this will all work on Windows vs `d3d11va` or a hypothetical `d3d12va`, and given that we're unlikely to see support for older codecs, we'd probably need a multi-hwdec fallback mechanism, which doesn't exist today. The priority right now is just getting the code finalised and merged.

**Comment by DeadSix27** [comment]
Did Lynne post the patch for ffmpeg public ?

**Comment by DeadSix27** [comment]
> > Did Lynne post the patch for ffmpeg public ?
> 
> https://github.com/cyanreg/FFmpeg/tree/vulkan

Of course I checked every site but his branches, last issue I encountered is:
```
./libavutil/vulkan.c:338:27: error: 'VK_QUERY_TYPE_VIDEO_ENCODE_BITSTREAM_BUFFER_RANGE_KHR' undeclared (first use in this function)
  338 |         if (query_type == VK_QUERY_TYPE_VIDEO_ENCODE_BITSTREAM_BUFFER_RANGE_KHR) {
  ```
  any clue on that? I looked through the web to see where it's defined but couldn't really find a proper source for it other than some mentions in the vulkan docs.

**Comment by DeadSix27** [comment]
> > > > Did Lynne post the patch for ffmpeg public ?
> > > 
> > > 
> > > https://github.com/cyanreg/FFmpeg/tree/vulkan
> > 
> > 
> > Of course I checked every site but his branches, last issue I encountered is:
> > ```
> > ./libavutil/vulkan.c:338:27: error: 'VK_QUERY_TYPE_VIDEO_ENCODE_BITSTREAM_BUFFER_RANGE_KHR' undeclared (first use in this function)
> >   338 |         if (query_type == VK_QUERY_TYPE_VIDEO_ENCODE_BITSTREAM_BUFFER_RANGE_KHR) {
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > any clue on that? I looked through the web to see where it's defined but couldn't really find a proper source for it other than some mentions in the vulkan docs.
> 
> I'm not vulkan expert either, so I removed this IF statement and let it pass the compiler checks, which shouldn't cause a problem since vulkan-encode is disabled by default.

Ran into more issues down the line, guess ill just wait until phillip and the others finished the rest before testing it.

(Issues later on: 
```
../video/out/vulkan/context.c:350:10: error: 'struct pl_vulkan_params' has no member named 'enable_video_decode_queue'
  350 |         .enable_video_decode_queue = true,

../video/out/hwdec/hwdec_vulkan.c: In function 'vulkan_init':
../video/out/hwdec/hwdec_vulkan.c:96:57: error: 'const struct pl_vulkan_t' has no member named 'queue_video_decode'
   96 |     device_hwctx->queue_family_decode_index = vk->vulkan->queue_video_decode.index;
      |                                                         ^~
../video/out/hwdec/hwdec_vulkan.c:97:48: error: 'const struct pl_vulkan_t' has no member named 'queue_video_decode'
   97 |     device_hwctx->nb_decode_queues = vk->vulkan->queue_video_decode.count;

```
)

**Comment by philipl** [comment]
> VK_QUERY_TYPE_VIDEO_ENCODE_BITSTREAM_BUFFER_RANGE_KHR

You need new enough Vulkan headers. The `VK_KHR_video_encode_queue` extension was added in 1.2.175, although it's unclear whether all packages of the headers include provisional extensions. In practice you want very recent (newer than Dec 2022) headers to cover all the extensions that the ffmpeg code wants.

**Comment by DeadSix27** [comment]
> > VK_QUERY_TYPE_VIDEO_ENCODE_BITSTREAM_BUFFER_RANGE_KHR
> 
> You need new enough Vulkan headers.
I'm using git master on all of vulkan though thats the thing.. 1.3.245 to be precise.

**Comment by philipl** [comment]
> > You need new enough Vulkan headers.
> 
> I'm using git master on all of vulkan though thats the thing.. 1.3.245 to be precise.

Maybe you need to configure something to expose provisional extensions?

**Comment by philipl** [comment]
```c
#ifdef VK_ENABLE_BETA_EXTENSIONS
    VK_QUERY_TYPE_VIDEO_ENCODE_BITSTREAM_BUFFER_RANGE_KHR = 1000299000,
#endif

```

**Comment by DeadSix27** [comment]
> Maybe you need to configure something to expose provisional extensions?

Wouldn't know where, I'm just building with the normal latest commits from vulkan loader/header, not like I need a newer patch or branch for those too right, maybe I'm missing something though.. but afaik there's no special build flag in their cmakelists and ffmpeg sets the beta define itself, so not sure...

Where exactly is `VK_QUERY_TYPE_VIDEO_ENCODE_BITSTREAM_BUFFER_RANGE_KHR` defined? Is it dynamically written by python code?

**Comment by philipl** [comment]
> Where exactly is `VK_QUERY_TYPE_VIDEO_ENCODE_BITSTREAM_BUFFER_RANGE_KHR` defined? Is it dynamically written by python code?

I don't know how the header build works - never tried running it. But I'd guess there's some build time option that controls whether the beta extensions are incorporated or not.

**Comment by philipl** [comment]
At this stage, we're basically done. Once the ffmpeg code is merged and we have a final version number to check against, we should be able to merge.

Currently requires one extra [patch](https://github.com/philipl/FFmpeg/commit/b1ee7d70034c8ebacfb55e1e6744835bb1a5d87b) on top of @cyanreg's branch to handle queue assignment.

**Comment by philipl** [comment]
Good work Intel.

**Comment by philipl** [comment]
Updated to skip requesting features/extensions that libplacebo now handles itself.

**Comment by Obegg** [comment]
(Might be a stupid question) Shouldn't this PR add changes to DOCS? (changelog, notes)

**Comment by DeadSix27** [comment]
Tested it with your latest revision, I get only green screen on nonHDR and HDR content, as well as with vulkan-copy and vulkan hwdec, I used cyanregs latest code with your patch and latest mpv

Debug txt: https://0x0.st/HKuu.txt

**Comment by philipl** [comment]
I don't think anyone really knows what to expect from Windows, which also means we don't know what the minimum driver version is.  You should install the latest libplacebo 273+ for sure though.

Try running with `--gpu-debug=yes` and see what it complains about. You will need to apply https://code.videolan.org/videolan/libplacebo/-/merge_requests/404 to libplacebo to avoid it failing on a false positive.

**Comment by philipl** [comment]
That's feedback for @cyanreg, I expect. This doesn't look like it is interop specific, given it looks wrong in the same way with `vulkan-copy`.

**Comment by DeadSix27** [comment]
> Try running with `--gpu-debug=yes` and see what it complains about
There: https://0x0.st/HKuG.txt
with latest placebo, all patches etc, I can't see any issue, idk..

**Comment by kasper93** [comment]
@DeadSix27: Could you make an issue about `restrict` keyword? With compiler information you are using. Thanks.

**Comment by DeadSix27** [comment]
> @DeadSix27: Could you make an issue about `restrict` keyword? With compiler information you are using. Thanks.

On the videolan repo?

**Comment by kasper93** [comment]
> On the videolan repo?

Yes, can also be on github mirror.

**Comment by philipl** [comment]
Updated to do decode queue discovery. There is no longer a dependency on any additional ffmpeg or libplacebo patches.

Works with @cyanreg's vulkan branch and libplacebo master.

**Comment by syncrond** [comment]
Hey, first of all thanks for all the work!

I did some testing of the vulkan video decode WIP recently on Intel with a `Intel(R) Iris(R) Plus Graphics 640 (Kaby Lake GT3e)` IGP.
Running on an Ubuntu 23.04 with the latest mesa 23.1.

Video decoding works for me for H264 without issues using the ANV env variable `ANV_VIDEO_DECODE`.
```
ANV_VIDEO_DECODE=1 mpv -v --vo=gpu-next --gpu-api=vulkan --hwdec=vulkan ~/ISAM\ Live\ HD.mp4
```
For H265 there is an open mesa MR https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/22202 which I haven't tested.

For the `hwmap` and `hwupload` examples mentioned in the comments above they show a picture but have errors with the filter graph:

`hwupload` format filter:
```
ANV_VIDEO_DECODE=1 mpv -v -v --vo=gpu-next --gpu-api=vulkan --hwdec=vaapi --vf=format=fmt=vulkan  ~/ISAM\ Live\ HD.mp4
```
results in:
```
[autoconvert] HW-uploading to vulkan
[hwupload] upload vaapi -> vulkan[nv12]
[hwupload] failed to upload frame
[vf] Disabling filter format.00 because it has failed.
```


`hwupload` lavfi filter:
```
ANV_VIDEO_DECODE=1 mpv -v --vo=gpu-next --gpu-api=vulkan --hwdec=vaapi --vd-lavc-dr=no --vf=lavfi=[hwupload,vflip_vulkan]:hwdec_interop=vulkan ~/ISAM\ Live\ HD.mp4
```
results in:
```
[lavfi] Configuring hwdec_interop=vulkan for filter graph: hwupload,vflip_vulkan
[ffmpeg] Impossible to convert between the formats supported by the filter 'mpv_src_in0' and the filter 'auto_scale_0'
[lavfi] failed to configure the filter graph
[vf] Disabling filter lavfi.00 because it has failed.
```

`hwmap:`
```
ANV_VIDEO_DECODE=1 mpv -v --vo=gpu-next --gpu-api=vulkan --hwdec=vaapi --vd-lavc-dr=no --vf=lavfi=[hwmap,vflip_vulkan]:hwdec_interop=vulkan ~/ISAM\ Live\ HD.mp4
```
results in:
```
[lavfi] Configuring hwdec_interop=vulkan for filter graph: hwmap,vflip_vulkan
[ffmpeg] Parsed_vflip_vulkan_1: Vulkan filtering requires that the VK_EXT_descriptor_buffer extension is supported!
[ffmpeg] Parsed_vflip_vulkan_1: Failed to configure output pad on Parsed_vflip_vulkan_1
[lavfi] failed to configure the filter graph
[vf] Disabling filter lavfi.00 because it has failed.
```

For the missing descriptor buffer extension there is also an open mesa MR: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/21645

**Comment by syncrond** [comment]
@philipl I added `--no-config` and captured a logfile: https://0x0.st/HNA7.log
I don't see anything about `hwmap` in the logfile but there some lines with `hwupload`?!

**Comment by philipl** [comment]
Weird. That works for me here. The main difference is I'm using X11 and you're using Wayland, but that shouldn't affect this part of the process. You won't see `hwmap` mentioned because `hwupload` here refers the mpv concept and not the ffmpeg one. Internally, it's doing a map.

**Comment by syncrond** [comment]
I got the format filter working!
After some debugging I found that `av_hwframe_map` returns ENOSYS in `mp_av_pool_image_hw_map`.
Then I looked at the ffmpeg code and figured the mapping code is behind a `#if CONFIG_LIBDRM` to map from vaapi. Looking at my ffmpeg compile options I found I missed enabling libdrm.
After enabling that mpv can successfully `hwmap` and the format filter works!

**Comment by syncrond** [comment]
A couple of days ago the H265 decoder was merged for Intel ANV Mesa Vulkan driver,
so I just gave it a shot.
I tried a couple of videos but non worked. I used the repos/branches (mpv, libplacebo, ffmpeg) from today.
The first errors in the log:
```
[   6.961][d][ffmpeg/video] hevc: Decoding frame, 626 bytes, 1 slices
[   6.961][e][ffmpeg/video] hevc: Unable to perform query: VK_ERROR_DEVICE_LOST!
[   6.961][e][ffmpeg/video] hevc: hardware accelerator failed to decode picture
[   6.961][w][vd] Error while decoding frame (hardware decoding)!
[   6.961][w][ffmpeg] av_log callback called with bad parameters (NULL AVClass).
[   6.961][w][ffmpeg] This is a bug in one of Libav/FFmpeg libraries used.
[   6.961][e][ffmpeg] Unable to submit command buffer: VK_ERROR_DEVICE_LOST
[   6.961][e][vo/gpu-next/libplacebo] vkQueueSubmit2: VK_ERROR_DEVICE_LOST (../src/vulkan/command.c:358)
[   6.961][e][ffmpeg/video] hevc: get_buffer() failed
[   6.961][w][ffmpeg/video] hevc: Error parsing NAL unit #0.
```
Full log is here: https://0x0.st/Hqc8.log
I tried for example this video: https://test-videos.co.uk/vids/bigbuckbunny/mp4/h265/1080/Big_Buck_Bunny_1080_10s_10MB.mp4

**Comment by philipl** [comment]
VK_ERROR_DEVICE_LOST is usually a sign of a very unhappy driver, which at this stage probably means a driver bug. I'm not sure how meaningful it is the try out at this stage, but if you're serious about validating the Intel hevc behaviour, you'd need to start with pure ffmpeg testing and discuss with @cyanreg (most easily on #ffmpeg-devel).

**Comment by syncrond** [comment]
Alright thank you, yeah I know its in an early stage but h264 worked fine so I gave h265 a shot.
I quickly tested with ffmpeg using:
`./ffmpeg_g -init_hw_device "vulkan=vk:0,debug=1" -hwaccel vulkan -hwaccel_output_format vulkan -i ~/Downloads/Big_Buck_Bunny_1080_10s_10MB.mp4 -loglevel debug -filter_hw_device vk -vf hwdownload,format=nv12 -c:v rawvideo -an -y OUT.nut`

but it results in the same errors, here's the logfile: https://0x0.st/HqAy.log if @cyanreg is interested.
Otherwise I might wait until the ffmpeg/mpv code is merged.

**Pull request merged**
